lib/ts/recipe/anomalydetection/types.ts

```diff
@@ -0,0 +1,113 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import OverrideableBuilder from "supertokens-js-override";
+import type { BaseRequest, BaseResponse } from "../../framework";
+import NormalisedURLDomain from "../../normalisedURLDomain";
+import NormalisedURLPath from "../../normalisedURLPath";
+
+export type TypeInput = {
+    basePath?: string;
+    domain?: string;
+    mode?: "BLOCK_ACCESS" | "REPORT_ONLY";
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    domain: NormalisedURLDomain;
+    basePath: NormalisedURLPath;
+    mode: "BLOCK_ACCESS" | "REPORT_ONLY";
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+type CheckAnomalyBaseRequest = {
+    action: "SIGN_UP" | "SIGN_IN" | "REFRESH_TOKEN";
+} & (
+    | {
+          status: "OK";
+          userId: string;
+          tenantId: string;
+      }
+    | {
+          status:
+              | "LINKING_TO_SESSION_USER_FAILED"
+              | "WRONG_CREDENTIALS_ERROR"
+              | "SIGN_IN_NOT_ALLOWED"
+              | "EMAIL_ALREADY_EXISTS_ERROR";
+          userId?: string;
+          tenantId?: string;
+      }
+    | {}
+);
+
+export type RecipeInterface = {
+    checkAnomaly: (
+        input: {
+            ipAddress?: string;
+            headers?: DeviceHeaders;
+        } & CheckAnomalyBaseRequest
+    ) => Promise<
+        | {
+              status: "OK";
+          }
+        | {
+              status: "ANOMALY_DETECTED";
+          }
+    >;
+    checkAnomalyWithRequest: (
```

Comment:
why do we need this function if we already have checkAnomaly function?

Reply:
the idea with the checkAnomaly function is to be reused, the sdk exposed function use it without the need of request object. this checkAnomalyWithRequest has helpers to extract the data from request object.

Reply:
I don't think this needs to be a recipe function. The other recipes can use a utils or something that will return the ip + the headers from the request object and then you can pass that into the other recipe function (that does not take a request) -> overall cleaner

Reply:
is true but it will increase the code changes on the other recipes, we are exposing the checkAnomaly that recipies can use. Also utils like getIpAddressFromRequest and getDeviceHeadersFromRequest